### ISSUE TITLE
Errors not reported on Cloud Run

### ISSUE THREAD

**Issue by DominicKramer** [googleapis/nodejs-logging-winston#354]
As reported in issue #32:

- At present I have configured nodejs-logging-winston for Google Cloud Run service and the errors don't appears in stackdriver error console. I have setup the serviceContext but it didnt work.

**Comment by aaaguirrep** [comment]
Hi, any status about the bug?

**Comment by JustinBeckwith** [comment]
Nope, not yet

**Comment by distractedCoding** [comment]
Hey! Any news about the bug ?

**Comment by DominicKramer** [comment]
Logs appear to be working now on Cloud Run and so I am closing this.

Here is the example I used to confirm that logging works.  In a new project directory, I created the files:

**package.json**
```json
{
  "name": "knative-serving-helloworld",
  "version": "1.0.0",
  "description": "Simple hello world sample in Node",
  "main": "index.js",
  "scripts": {
    "start": "node index.js"
  },
  "author": "",
  "license": "Apache-2.0",
  "dependencies": {
    "@google-cloud/logging-winston": "^3.0.0",
    "express": "^4.16.4",
    "winston": "^3.2.1"
  }
}
```

**index.js:**
```js
const winston = require('winston');
const {LoggingWinston} = require('@google-cloud/logging-winston');
const loggingWinston = new LoggingWinston();

const logger = winston.createLogger({
  level: 'info',
  transports: [
    new winston.transports.Console(),
    // Add Stackdriver Logging
    loggingWinston,
  ],
});

const express = require('express');
const app = express();

app.get('/', (req, res) => {
  logger.info('Received / request at ' + new Date());
  const target = process.env.TARGET || 'World';
  res.send(`Hello ${target}!`);
});

const port = process.env.PORT || 8080;
app.listen(port, () => {
  console.log('Hello world listening on port', port);
});
```

**Dockerfile:**
```docker
# Use the official lightweight Node.js 12 image.
# https://hub.docker.com/_/node
FROM node:12-slim

# Create and change to the app directory.
WORKDIR /usr/src/app

# Copy application dependency manifests to the container image.
# A wildcard is used to ensure both package.json AND package-lock.json are copied.
# Copying this separately prevents re-running npm install on every code change.
COPY package*.json ./

# Install production dependencies.
RUN npm install --only=production

# Copy local code to the container image.
COPY . ./

# Run the web service on container startup.
CMD [ "npm", "start" ]
```

From that directory, I built the application's container image:
```
gcloud builds submit --tag gcr.io/my-projectid/helloworld
```
and then deployed to Cloud Run:
```
gcloud run deploy --image gcr.io/my-projectid/helloworld --platform=managed
```

After accessing the running application, the logs appear in the the [Logs Viewer](https://pantheon.corp.google.com/logs/viewer) under "Cloud Run Revision" > "helloworld".

**Comment by boredland** [comment]
I think you missunderstood the issue. This was not about the logging tool, but about the "Error Reporting". As it is described in the Issue description, I can reproduce it.

**Comment by boredland** [comment]
To reiterate: 
Just change your line `  logger.info('Received / request at ' + new Date());` to error. And specify a service context:
```
const loggingWinston = new LoggingWinston({
serviceContext: {
    service: 'my-service', // required to report logged errors
                        // to the Google Cloud Error Reporting
                        // console
    version: 'my-version'
}
});
```
After that, an error should be reported inside your Stackdriver "Error Reporting" Dashboard.

**Comment by stianlp** [comment]
Same issue here. Error logs aren't showing up even with serviceContext. Logs are working fine.

**Comment by wissam-launchtrip** [comment]
Any update on this?
Issue confirmed: error reporting does not work.

**Comment by DominicKramer** [comment]
We are still looking at this.  Thanks.

**Comment by soldair** [comment]
This is an issue with how the error reporting service ingests logs from cloud run as far as we can guess. All of the data you log ends up in the log data in cloud logs but not in error reporting.

If you are a google cloud customer and are experiencing this issue feel free to open a support ticket with error reporting so the appropriate team can resolve the issue. Also consider that instead of logging errors by proxy via the logs you may consider using the error reporting api directly.

**Comment by bcoe** [comment]
@soldair for these sorts of things where it needs to be fixed upstream, we can toss the `external` label on. It sounds like there are potentially some plans in the works to help address this issue, for what it's worth, but no clear timelines yet.

**Comment by wissam-launchtrip** [comment]
We're experiencing this issue in GKE as well. FYI.

**Comment by soldair** [comment]
because we handle serviceContexts etc in nodejs-logging im closing this issue.

https://github.com/googleapis/nodejs-logging/blob/master/src/metadata.ts#L174

if serviceContext is being set correctly or manually i would expect the errors to show up but that should be brought up as an issue with the cloud error reporting team or at least be a bug/docs in nodejs-logging

**Comment by aaron-ikb** [comment]
@soldair the metadata link is really helpful. For my specific problem, I couldn't find a closed or open issue, or any stackoverflow answer. But your link, the discussion here, helps point me in the right direction.

- The problem: my Cloud Run logs are being logged as `gce_instance` logs.
- The hypothesis: Cloud Run is not a detectable service context/resource, so it incorrectly selects `gce_instance` as the "resource type."
- The questions: Is my hypothesis correct? Could you point me to some examples of how to manually set the `serviceContext`/"resource type" to work for Cloud Run? I guess it's not the logging team's job to support detecting the Cloud Run environment, so I'll just have to jump over to https://github.com/googleapis/google-auth-library-nodejs to file an issue?

I could file any issue if needed.

**Comment by peterwiebe** [comment]
@aaron-ikb did you get anywhere with setting the correct resource type? Logs Explorer is showing it as `gce_instance` for me too 😞

**Comment by aaron-ikb** [comment]
@peterwiebe unfortunately, I did not. For now we've switched to GAE flexible for our solution so we have not pursued this issue further and it seems like no one has responded either 🤷

**Comment by peterwiebe** [comment]
@aaron-ikb I managed to figure out both the logging and the error reporting:
```
const loggingWinston = new LoggingWinston({
  resource: {
    type: "cloud_run_revision",
    labels: {
      configuration_name: process.env.K_CONFIGURATION,
      location: process.env.K_LOCATION,
      revision_name: process.env.K_REVISION,
      service_name: process.env.K_SERVICE,
    },
  },
  serviceContext: {
    service: process.env.K_SERVICE,
    version: process.env.K_REVISION,
  },
})
```
you need to manually define the resource property to get the structured winston logs to show up in the Cloud Run Logs tab. All of the environment variables are defined by the Cloud Run runtime contract _except_ the one you really need which is for `location`, I defined that variable manually based on where we deployed our service. (The Cloud Run logs filter logs based on resource.type and resource.labels.location.)

For the Error Reporting you need to of course have the `serviceContext` defined but when logging errors with winston you need to make sure that they make use of `new Error()` otherwise they don't land in Error Reporting:
```
logger.error(new Error('Something broke'))
```
or
```
logger.error('Error message', new Error('Something else broke'))
```

**Comment by aaron-ikb** [comment]
@peterwiebe the detailed instructions are really helpful... especially all the tips. Just to clarify, you had to manually define `K_LOCATION`?

It is unfortunate that this needs to be manually set, because it couples the code to the infrastructure, e.g. if you decided to deploy to App Engine, you would need to remove this code.

**Comment by peterwiebe** [comment]
@aaron-ikb I defined `K_LOCATION` in the environment variables config in the settings of the Cloud Run project (using the cloud console interface). It is set to the location you use when deploying. 

I agree that this is unfortunate and wonder why it is set automatically.

**Comment by Daniel-Sanche** [comment]
I was requested to re-open this issue. I'm assigning @nicoleczhu as she has the most experience with this repo, but from my understanding:
- Cloud Run is not yet officially supported by the repo, so for now the resource will have to be manually set instead of auto-detected. We are working on improving support, and hope to have it in place next quarter
- The Cloud Logging libraries don't currently automatically send errors to the Error Reporting dashboard. For now, there is a [different library for this](https://github.com/googleapis/nodejs-error-reporting), but we may combine them at some point in the future

Let me know if I missed anything Nicole

**Comment by Daniel-Sanche** [comment]
(@nicoleczhu, realized I forgot to actually assign you. Let me know what you think)

**Comment by nicoleczhu** [comment]
Hi folks, echoing Dan's pts above. I'm working on populating GCR fields Q2. You can track that [here](https://github.com/googleapis/nodejs-logging/issues/959) and [here](https://github.com/googleapis/nodejs-logging/issues/1006).

**Comment by nicoleczhu** [comment]
Cloud run is in, will be released with [9.2.0 here](https://github.com/googleapis/nodejs-logging/pull/1025#pullrequestreview-629178260). Let me know if anything doesn't work as expected. Thanks all

**Comment by bhr** [comment]
@nicoleczhu I'm still facing issues automatically setting the resource for my services that run on Cloud Run. 
Logs do appear in Cloud Logging, but resource tags are still missing.

Any pointers what I'm doing wrong would be much appreciated.

**package.json**
```
  ...
  "dependencies": {
    "@google-cloud/logging": "^9.5.5",
    "@google-cloud/logging-winston": "^4.1.0",
    "@google-cloud/opentelemetry-cloud-trace-exporter": "^0.10.0",
    "@opentelemetry/api": "^1.0.1",
    "@opentelemetry/core": "^0.22.0",
    "@opentelemetry/instrumentation-express": "^0.22.0",
    "@opentelemetry/instrumentation-http": "^0.22.0",
    "@opentelemetry/node": "^0.21.0",
    "@opentelemetry/tracing": "^0.21.0",
    "@types/node": "^14.14.12",
    "gaxios": "^4.3.0",
    "winston": "^3.3.3"
  },
```

**logging.ts**
```
import winston from 'winston';
import { LoggingWinston } from '@google-cloud/logging-winston';

const isProduction = process.env.NODE_ENV === 'production';
const transports = isProduction
  ? [new LoggingWinston()]
  : [new winston.transports.Console({ level: 'debug', format: winston.format.combine(winston.format.colorize(), winston.format.simple()) })];

export const logger = winston.createLogger({
  level: 'debug',
  transports,
});

console.log = (...args: any[]) => {
  return logger.info.apply(logger, [...args] as any);
};

console.error = (...args: any[]) => {
  return logger.error.apply(logger, [...args] as any);
};

console.warn = (...args: any[]) => {
  return logger.warn.apply(logger, [...args] as any);
};

console.info = (...args: any[]) => {
  return logger.info.apply(logger, [...args] as any);
};

console.debug = (...args: any[]) => {
  return logger.debug.apply(logger, [...args] as any);
};
```
